app/scripts/lib/metaRPCClient.js

```diff
@@ -0,0 +1,77 @@
+import { EventEmitter } from 'events';
+import { EthereumRpcError } from 'eth-rpc-errors';
+import createRandomId from '../../../shared/modules/random-id';
+
+class MetaRPCClient {
+  constructor(connectionStream) {
+    this.connectionStream = connectionStream;
+    this.notificationChannel = new EventEmitter();
+    this.requests = new Map();
+    this.connectionStream.on('data', this.handleResponse.bind(this));
+    this.connectionStream.on('end', this.close.bind(this));
+    // eslint-disable-next-line no-constructor-return
+    return new Proxy(this, {
+      get: (object, property) => {
+        if (object[property]) {
+          return object[property];
+        }
+        return (...p) => {
+          const cb = p[p.length - 1];
+          const params = p.slice(0, -1);
+          const id = createRandomId();
+
+          object.requests.set(id, cb);
+          object.connectionStream.write({
+            method: property,
+            params,
+            id,
+          });
+        };
+      },
+    });
+  }
+
+  onNotification(handler) {
+    this.notificationChannel.addListener('notification', (data) => {
+      handler(data);
+    });
+  }
+
+  close() {
+    this.notificationChannel.removeAllListeners();
+  }
+
+  handleResponse(data) {
+    const { id, result, error, method, params } = data;
+    const cb = this.requests.get(id);
+
+    if (method && params && id) {
+      // dont handle server-side to client-side requests
```

Comment:
Should we throw in this case? This should be unreachable, if I understand this correctly.

Reply:
one of the reasons I have this is because calling `write` postMessage/websocket/portStream doesn't fire `data` on its own stream, but my testStream does and ends up here.

Reply:
I played around with this today but didn't quite manage to get it to work. We can improve this later if we get around to it - I don't see any reason to block on it.